Source/WebCore/css/CSSCounterStyleRule.cpp

```diff
@@ -220,104 +160,125 @@ void CSSCounterStyleRule::reattach(StyleRuleBase& rule)
     m_counterStyleRule = downcast<StyleRuleCounterStyle>(rule);
 }
 
+RefPtr<CSSValue> CSSCounterStyleRule::CSSValueFromText(CSSPropertyID propertyID, const String& valueText)
+{
+    auto tokenizer = CSSTokenizer(valueText);
+    auto tokenRange = tokenizer.tokenRange();
+    return CSSPropertyParser::parseCounterStyleDescriptor(propertyID, tokenRange, parserContext());
+}
+
 // https://drafts.csswg.org/css-counter-styles-3/#dom-csscounterstylerule-name
 void CSSCounterStyleRule::setName(const String& text)
 {
     auto tokenizer = CSSTokenizer(text);
     auto tokenRange = tokenizer.tokenRange();
     auto name = CSSPropertyParserHelpers::consumeCounterStyleNameInPrelude(tokenRange);
-    if (name.isNull() || name == m_counterStyleRule->name())
+    if (!name)
+        return;
+    if (!mutableDescriptors().setName(WTFMove(name)))
         return;
-
-    CSSStyleSheet::RuleMutationScope mutationScope(this);
-    m_counterStyleRule->setName(name);
-    m_counterStyleRule->mutableDescriptors().m_name = name;
-}
-
-bool CSSCounterStyleRule::setterInternal(CSSPropertyID propertyID, const String& valueText)
-{
-    auto tokenizer = CSSTokenizer(valueText);
-    auto tokenRange = tokenizer.tokenRange();
-    auto newValue = CSSPropertyParser::parseCounterStyleDescriptor(propertyID, tokenRange, parserContext());
-    if (m_counterStyleRule->newValueInvalidOrEqual(propertyID, newValue))
-        return false;
-
     CSSStyleSheet::RuleMutationScope mutationScope(this);
-    m_counterStyleRule->mutableProperties().setProperty(propertyID, WTFMove(newValue));
-    return true;
 }
 
 void CSSCounterStyleRule::setSystem(const String& text)
 {
-    if (!setterInternal(CSSPropertySystem, text))
+    auto systemValue = CSSValueFromText(CSSPropertySystem, text);
+    if (!systemValue)
         return;
-    const auto& properties = m_counterStyleRule->properties();
-    auto systemValue = properties.getPropertyCSSValue(CSSPropertySystem);
     auto system = toCounterStyleSystemEnum(systemValue.get());
-    auto systemData = extractDataFromSystemDescriptor(properties, system);
-    m_counterStyleRule->mutableDescriptors().m_fixedSystemFirstSymbolValue = systemData.second;
-    m_counterStyleRule->mutableDescriptors().m_extendsName = systemData.first;
+    // If the attribute being set is `system`, and the new value would change the algorithm used, do nothing
+    // and abort these steps.
+    // (It's okay to change an aspect of the algorithm, like the first symbol value of a `fixed` system.)
+    // https://www.w3.org/TR/css-counter-styles-3/#the-csscounterstylerule-interface
+    auto systemData = extractSystemDataFromCSSValue(WTFMove(systemValue), system);
+    if (!mutableDescriptors().setSystemData(WTFMove(systemData)))
+        return;
+    CSSStyleSheet::RuleMutationScope mutationScope(this);
```

Comment:
RuleMutationScope should be scoped over the mutation. That is the constructor should run before the mutation and destructor after (here and elsewhere).